### PR TITLE
Enabling custom named connector actions and support PUT/DELETE action

### LINKED ISSUE #4569: [RFC] Enable Custom Named Connector Actions and Support PUT/DELETE HTTP Methods
This RFC proposes enhancing the ML Commons connector framework to support custom named actions and additional HTTP methods (PUT/DELETE), enabling more flexible and powerful connector configurations.

Currently, the ML Commons connector framework has the following limitations:

- **Limited Action Types**: Connectors only support a fixed set of action types (PREDICT, EXECUTE), which restricts flexibility.

- **No Support for Multiple Actions of Same Type**: Users cannot define multiple actions with the same action_type. For example, a connector cannot have three different EXECUTE actions for write, search, and delete operations.

- **Limited HTTP Methods**: Only `GET` and `POST` methods are supported, preventing integration with REST APIs that require `PUT` or `DELETE` operations.

# Proposed Solution
1. Add name Field to Connector Actions

Introduce an optional `name` field in the connector action definition to uniquely identify actions, allowing multiple actions of the same type.

2. Support PUT/DELETE HTTP Methods

Extend the supported HTTP methods to include `PUT` and `DELETE` for broader REST API compatibility.

## API Changes

Connector Creation Example
```
POST _plugins/_ml/connectors/_create
{
  "name": "test connector",
  "description": "Connector with multiple named actions",
  "version": 1,
  "protocol": "http",
  "parameters": {
    "endpoint": "https://api.example.com"
  },
  "credential": {
    "api_key": "..."
  },
  "actions": [
    {
      "action_type": "execute",
      "name": "write_data",
      "method": "POST",
      "url": "{{endpoint}}/data",
      "headers": {
        "Content-Type": "application/json"
      },
      "request_body": "${parameters.input}"
    },
    {
      "action_type": "execute",
      "name": "search_data",
      "method": "GET",
      "url": "{{endpoint}}/data/_search",
      "headers": {
        "Content-Type": "application/json"
      },
      "request_body": "${parameters.input}"
    },
    {
      "action_type": "execute",
      "name": "update_data",
      "method": "PUT",
      "url": "{{endpoint}}/data/${parameters.id}",
      "headers": {
        "Content-Type": "application/json"
      },
      "request_body": "${parameters.input}"
    },
    {
      "action_type": "execute",
      "name": "delete_data",
      "method": "DELETE",
      "url": "{{endpoint}}/data/${parameters.id}",
      "headers": {
        "Content-Type": "application/json"
      }
    }
  ]
}
```

Invoking Named Actions

```
POST _plugins/_ml/models/<model_id>/_execute
{
  "action_name": "write_data",
  "parameters": {
    "input": "{\"key\": \"value\"}"
  }
}
```

## Backward Compatibility

- The name field is optional; existing connectors without named actions will continue to work
- Default behavior (resolving by action_type) remains unchanged when action_name is not specified
- No breaking changes to existing APIs

### PR DESCRIPTION
**Pull request by b4sjoo**
### Description

Currently the connector only supports limited action type like `PREDICT` , `EXECUTE`. That can't support multiple actions with same type like `EXECUTE`. For example one connector may contains three `EXECUTE` actions:  write data, search data, delete data. To support this, we will add a new field `name` in connector's action definition.

This PR enables custom named connector actions and supports PUT/DELETE action, which will be the prerequisite of remote agentic memory container.

Example
```
POST _plugins/_ml/connectors/_create
{
  "name": "test connector",
  "description": "This is a test connector",
  "version": 1,
  "protocol": "http",
  "parameters": {
    ...
  },
  "credential": {
    ...
  },
  "actions": [
    {
      "action_type": "execute",
      "name": "write_data",
      "method": "post",
      "url": "{{endpoint}}/_write",
      "headers": {
        ...
      },
      "request_body": "${parameters.input}"
    },
    {
      "action_type": "execute",
      "name": "search_data",
      "method": "get",
      "url": "{{endpoint}}/_search",
      "headers": {
        ...
      },
      "request_body": "${parameters.input}"
    },
```

### Related Issues
Resolves #4569

### Check List
- [ ] New functionality includes testing.
- [ ] New functionality has been documented.
- [ ] API changes companion pull request [created](https://github.com/opensearch-project/opensearch-api-specification/blob/main/DEVELOPER_GUIDE.md).
- [ ] Commits are signed per the DCO using `--signoff`.
- [ ] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose).

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/ml-commons/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

## Summary by CodeRabbit

* **New Features**
  * Added support for custom action naming within connectors
  * Enabled NDJSON (newline-delimited JSON) payload format support
  * Added PUT and DELETE HTTP method support for connector operations

* **Bug Fixes**
  * Improved connector action lookup logic with enhanced null checks
  * Fixed payload validation to accept both JSON and NDJSON formats

* **Tests**
  * Comprehensive test coverage for action lookup scenarios and edge cases
  * Validation tests for JSON and NDJSON format handling

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

### CODE REVIEW

File: common/src/main/java/org/opensearch/ml/common/connector/ConnectorAction.java [hunk 1]
```diff
@@ -97,6 +110,9 @@ public ConnectorAction(StreamInput input) throws IOException {
         this.requestBody = input.readOptionalString();
         this.preProcessFunction = input.readOptionalString();
         this.postProcessFunction = input.readOptionalString();
+        if (input.getVersion().onOrAfter(VERSION_3_4_0)) {
```
Comment (ylwu-amzn) [inline common/src/main/java/org/opensearch/ml/common/connector/ConnectorAction.java]: Should check with 3.5?
```suggestion
        if (input.getVersion().onOrAfter(VERSION_3_5_0)) {
```

File: common/src/main/java/org/opensearch/ml/common/connector/ConnectorAction.java [hunk 2]
```diff
@@ -113,6 +129,9 @@ public void writeTo(StreamOutput out) throws IOException {
         out.writeOptionalString(requestBody);
         out.writeOptionalString(preProcessFunction);
         out.writeOptionalString(postProcessFunction);
+        if (out.getVersion().onOrAfter(VERSION_3_4_0)) {
```
Comment (ylwu-amzn) [inline common/src/main/java/org/opensearch/ml/common/connector/ConnectorAction.java]: Should check with 3.5?

File: common/src/main/java/org/opensearch/ml/common/connector/ConnectorAction.java (hunk 1 above)
Comment (b4sjoo) [reply common/src/main/java/org/opensearch/ml/common/connector/ConnectorAction.java]: Done

File: common/src/main/java/org/opensearch/ml/common/connector/ConnectorAction.java (hunk 2 above)
Comment (b4sjoo) [reply common/src/main/java/org/opensearch/ml/common/connector/ConnectorAction.java]: Right, changed.

File: plugin/src/main/java/org/opensearch/ml/action/connector/ExecuteConnectorTransportAction.java [hunk 3]
```diff
@@ -73,14 +75,24 @@ public ExecuteConnectorTransportAction(
     protected void doExecute(Task task, ActionRequest request, ActionListener<MLTaskResponse> actionListener) {
         MLExecuteConnectorRequest executeConnectorRequest = MLExecuteConnectorRequest.fromActionRequest(request);
         String connectorId = executeConnectorRequest.getConnectorId();
+        if (executeConnectorRequest.getMlInput() == null) {
+            actionListener.onFailure(new IllegalArgumentException("MLInput cannot be null"));
+            return;
+        }
+
+        RemoteInferenceInputDataSet inputDataset = (RemoteInferenceInputDataSet) executeConnectorRequest.getMlInput().getInputDataset();
```
[the remark this answers is not in the document]
Comment (dhrubo-os) [reply plugin/src/main/java/org/opensearch/ml/action/connector/ExecuteConnectorTransportAction.java:83]: Can we address this?

**Comment by ylwu-amzn** [comment]
CI failed 
```
REPRODUCE WITH: ./gradlew ':opensearch-ml-plugin:integTest' --tests 'org.opensearch.ml.rest.RestConnectorToolIT.testConnectorToolInFlowAgent_WrongAction' -Dtests.seed=3F231DB1717F5444 -Dtests.security.manager=false -Dtests.locale=en-SG -Dtests.timezone=Europe/Belfast -Druntime.java=25
RestConnectorToolIT > testConnectorToolInFlowAgent_WrongAction STANDARD_ERROR
    REPRODUCE WITH: ./gradlew ':opensearch-ml-plugin:integTest' --tests 'org.opensearch.ml.rest.RestConnectorToolIT.testConnectorToolInFlowAgent_WrongAction' -Dtests.seed=3F231DB1717F5444 -Dtests.security.manager=false -Dtests.locale=en-SG -Dtests.timezone=Europe/Belfast -Druntime.java=25

RestConnectorToolIT > testConnectorToolInFlowAgent_WrongAction FAILED
    java.lang.AssertionError: 
    Expected: a string containing "no EXECUTE action found"
         but: was "method [POST], host [http://[::1]:33785], URI [/_plugins/_ml/agents/bw4B_JsBFGRGWsMrvU28/_execute], status line [HTTP/1.1 400 Bad Request]
    Warnings: [this request accesses system indices: [.plugins-ml-agent], but in a future major version, direct access to system indices will be prevented by default]
    {"status":400,"error":{"type":"IllegalArgumentException","reason":"Invalid Request","details":"Some parameter placeholder not filled in payload: messages"}}"
        at __randomizedtesting.SeedInfo.seed([3F231DB1717F5444:A9FCF83B06B07498]:0)
        at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:18)
        at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:6)
        at org.opensearch.ml.rest.RestConnectorToolIT.testConnectorToolInFlowAgent_WrongAction(RestConnectorToolIT.java:128)
```

**Comment by b4sjoo** [comment]
> CI failed
> 
> ```
> REPRODUCE WITH: ./gradlew ':opensearch-ml-plugin:integTest' --tests 'org.opensearch.ml.rest.RestConnectorToolIT.testConnectorToolInFlowAgent_WrongAction' -Dtests.seed=3F231DB1717F5444 -Dtests.security.manager=false -Dtests.locale=en-SG -Dtests.timezone=Europe/Belfast -Druntime.java=25
> RestConnectorToolIT > testConnectorToolInFlowAgent_WrongAction STANDARD_ERROR
>     REPRODUCE WITH: ./gradlew ':opensearch-ml-plugin:integTest' --tests 'org.opensearch.ml.rest.RestConnectorToolIT.testConnectorToolInFlowAgent_WrongAction' -Dtests.seed=3F231DB1717F5444 -Dtests.security.manager=false -Dtests.locale=en-SG -Dtests.timezone=Europe/Belfast -Druntime.java=25
> 
> RestConnectorToolIT > testConnectorToolInFlowAgent_WrongAction FAILED
>     java.lang.AssertionError: 
>     Expected: a string containing "no EXECUTE action found"
>          but: was "method [POST], host [http://[::1]:33785], URI [/_plugins/_ml/agents/bw4B_JsBFGRGWsMrvU28/_execute], status line [HTTP/1.1 400 Bad Request]
>     Warnings: [this request accesses system indices: [.plugins-ml-agent], but in a future major version, direct access to system indices will be prevented by default]
>     {"status":400,"error":{"type":"IllegalArgumentException","reason":"Invalid Request","details":"Some parameter placeholder not filled in payload: messages"}}"
>         at __randomizedtesting.SeedInfo.seed([3F231DB1717F5444:A9FCF83B06B07498]:0)
>         at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:18)
>         at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:6)
>         at org.opensearch.ml.rest.RestConnectorToolIT.testConnectorToolInFlowAgent_WrongAction(RestConnectorToolIT.java:128)
> ```

Fixed and CI passed

File: common/src/main/java/org/opensearch/ml/common/utils/StringUtils.java [hunk 4]
```diff
@@ -150,6 +150,40 @@ public static boolean isJson(String json) {
...
+        }
+
+        // First check if it's regular JSON
+        if (isJson(json)) {
+            return true;
+        }
+
+        // Check if it's NDJSON (newline-delimited JSON)
+        String[] lines = json.split("\\r?\\n");
+        if (lines.length == 0) {
```
Comment (dhrubo-os) [inline common/src/main/java/org/opensearch/ml/common/utils/StringUtils.java:172]: This seems like a dead code, since split() on non-blank string always returns at least one element

File: plugin/src/main/java/org/opensearch/ml/action/connector/ExecuteConnectorTransportAction.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (b4sjoo) [reply plugin/src/main/java/org/opensearch/ml/action/connector/ExecuteConnectorTransportAction.java:83]: Addressed in other PR

File: common/src/main/java/org/opensearch/ml/common/utils/StringUtils.java (hunk 4 above)
Comment (b4sjoo) [reply common/src/main/java/org/opensearch/ml/common/utils/StringUtils.java:172]: You are right, removed
